torch/_functorch/vmap.py

```diff
@@ -391,19 +398,114 @@ def vmap(
         sequences out of the box.
     """
     _check_randomness_arg(randomness)
+    if not (chunk_size is None or chunk_size > 0):
+        raise ValueError("vmap: `chunk_size` should be greater than 0.")
+
+    def _get_chunked_inputs(flat_args, flat_in_dims, batch_size, chunk_size):
+        split_idxs = (batch_size,)
+        if chunk_size is not None and chunk_size > 0:
+            n_chunks = n_chunks = batch_size // chunk_size
+            chunk_numels = [chunk_size] * n_chunks
+            # remainder chunk
+            chunk_numels.append(batch_size % chunk_size)
+            split_idxs = tuple(itertools.accumulate(chunk_numels))
+
+        flat_args_chunks = tuple(
+            t.tensor_split(split_idxs, dim=in_dim) if in_dim is not None else [t, ] * len(split_idxs)
+            for t, in_dim in zip(flat_args, flat_in_dims)
+        )
+
+        # transpose chunk dim and flatten structure
+        # chunks_flat_args is a list of flatten args
+        chunks_flat_args = zip(*flat_args_chunks)
+        return chunks_flat_args
 
     @functools.wraps(func)
     def wrapped(*args, **kwargs):
         lazy_load_decompositions()
         _check_out_dims_is_int_or_int_pytree(out_dims, func)
         batch_size, flat_in_dims, flat_args, args_spec = _process_batched_inputs(in_dims, args, func)
+
+        if chunk_size is not None:
+            chunks_flat_args = _get_chunked_inputs(flat_args, flat_in_dims, batch_size, chunk_size)
+            return _chunked_vmap(func, flat_in_dims, chunks_flat_args,
+                                 args_spec, out_dims, randomness, **kwargs)
+
+        # If chunk_size is not specified.
         return _flat_vmap(
             func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
         )
 
     return wrapped
 
 
+def _flatten_chunks_output(chunks_output_):
+    # chunks_output is a list of chunked outputs
+    # flatten chunked outputs:
+    flat_chunks_output = []
+    arg_spec = None
+    for output in chunks_output_:
+        flat_output, arg_specs = tree_flatten(output)
+        flat_chunks_output.append(flat_output)
+        if arg_spec is None:
+            arg_spec = arg_specs
+
+    # transpose chunk dim and flatten structure
+    # flat_output_chunks is flat list of chunks
+    flat_output_chunks = list(zip(*flat_chunks_output))
+    return flat_output_chunks, arg_spec
+
+
+def _concat_chunked_outputs(out_dims, arg_spec, flat_output_chunks):
+    # concat chunks on out_dim
+    flat_out_dims = _broadcast_to_and_flatten(out_dims, arg_spec)
```

Comment:
`_broadcast_to_and_flatten` can return None. In that case, out_dims is not compatible with arg_spec.

I'm not sure if this case is handled elsewhere in the code. But if it's not, we should raise a good error message. Would be good to add a test for vmap(f, out_dims, chunk_size=...) where out_dims is wrong

Reply:
This is already handled while computing `vmap` of the chunk.

Reply:
Awesome, thanks for checking